File: linkhive_app/lib/models/link.dart
class Link {
  final int? id;
  final int userId;
  final int? categoryId;
  final String url;
  final String? title;
  final LinkType type;
  final LinkStatus status;
  final Map<String, dynamic>? metadata;
  final bool sharedFlag;
  final bool pinnedFlag;
  final int orderIndex;
  final bool isDead;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final List<String>? tags;

  Link({
    this.id,
    required this.userId,
    this.categoryId,
    required this.url,
    this.title,
    this.type = LinkType.other,
    this.status = LinkStatus.none,
    this.metadata,
    this.sharedFlag = false,
    this.pinnedFlag = false,
    this.orderIndex = 0,
    this.isDead = false,
    this.createdAt,
    this.updatedAt,
    this.tags,
  });

  factory Link.fromJson(Map<String, dynamic> json) {
    return Link(
      id: json['id'],
      userId: json['user_id'],
      categoryId: json['category_id'],
      url: json['url'],
      title: json['title'],
      type: LinkType.values.firstWhere(
        (e) => e.toString().split('.').last == json['type'],
        orElse: () => LinkType.other,
      ),
      status: LinkStatus.values.firstWhere(
        (e) => e.toString().split('.').last == json['status'],
        orElse: () => LinkStatus.none,
      ),
      metadata: json['metadata'],
      sharedFlag: json['shared_flag'] ?? false,
      pinnedFlag: json['pinned_flag'] ?? false,
      orderIndex: json['order_index'] ?? 0,
      isDead: json['is_dead'] ?? false,
      createdAt: json['created_at'] != null 
          ? DateTime.parse(json['created_at']) 
          : null,
      updatedAt: json['updated_at'] != null 
          ? DateTime.parse(json['updated_at']) 
          : null,
      tags: json['tags'] != null 
          ? List<String>.from(json['tags'].map((tag) => tag['name'])) 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'category_id': categoryId,
      'url': url,
      'title': title,
      'type': type.toString().split('.').last,
      'status': status.toString().split('.').last,
      'metadata': metadata,
      'shared_flag': sharedFlag,
      'pinned_flag': pinnedFlag,
      'order_index': orderIndex,
      'is_dead': isDead,
    };
  }

  Link copyWith({
    int? id,
    int? userId,
    int? categoryId,
    String? url,
    String? title,
    LinkType? type,
    LinkStatus? status,
    Map<String, dynamic>? metadata,
    bool? sharedFlag,
    bool? pinnedFlag,
    int? orderIndex,
    bool? isDead,
    DateTime? createdAt,
    DateTime? updatedAt,
    List<String>? tags,
  }) {
    return Link(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      categoryId: categoryId ?? this.categoryId,
      url: url ?? this.url,
      title: title ?? this.title,
      type: type ?? this.type,
      status: status ?? this.status,
      metadata: metadata ?? this.metadata,
      sharedFlag: sharedFlag ?? this.sharedFlag,
      pinnedFlag: pinnedFlag ?? this.pinnedFlag,
      orderIndex: orderIndex ?? this.orderIndex,
      isDead: isDead ?? this.isDead,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      tags: tags ?? this.tags,
    );
  }
}

enum LinkType {
  job,
  reel,
  article,
  video,
  other,
}

enum LinkStatus {
  applied,
  notApplied,
  read,
  unread,
  none,
}

File: linkhive-backend/app/Models/Link.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Link extends Model
{
    use HasFactory;

    protected $fillable = [
        'user_id',
        'category_id',
        'url',
        'title',
        'type',
        'status',
        'metadata',
        'shared_flag',
        'pinned_flag',
        'order_index',
        'is_dead',
    ];

    protected $casts = [
        'metadata' => 'array',
        'shared_flag' => 'boolean',
        'pinned_flag' => 'boolean',
        'is_dead' => 'boolean',
    ];

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function category(): BelongsTo
    {
        return $this->belongsTo(Category::class);
    }

    public function tags(): BelongsToMany
    {
        return $this->belongsToMany(Tag::class, 'link_tag');
    }

    public function reminders(): HasMany
    {
        return $this->hasMany(Reminder::class);
    }

    public function sharedLinks(): HasMany
    {
        return $this->hasMany(SharedLink::class);
    }

    // Auto-detect link type based on URL
    public static function detectLinkType(string $url): string
    {
        $url = strtolower($url);
        
        if (str_contains($url, 'linkedin.com/jobs') || 
            str_contains($url, 'indeed.com') || 
            str_contains($url, 'naukri.com')) {
            return 'job';
        }
        
        if (str_contains($url, 'instagram.com/reel') || 
            str_contains($url, 'tiktok.com')) {
            return 'reel';
        }
        
        if (str_contains($url, 'youtube.com') || 
            str_contains($url, 'youtu.be') ||
            str_contains($url, 'vimeo.com')) {
            return 'video';
        }
        
        if (str_contains($url, 'medium.com') || 
            str_contains($url, 'dev.to') ||
            str_contains($url, '/article') ||
            str_contains($url, '/blog')) {
            return 'article';
        }
        
        return 'other';
    }
}